### DISCUSSION TITLE
How to test tRPC queries

### DISCUSSION THREAD

**Discussion by matthewd673** [stepci/stepci#105] [Q&A] [answered]
After reading [the documentation](https://docs.stepci.com/guides/testing-trpc.html) for tRPC support and debugging for several hours I haven't been able to get the feature to work.

Consider a basic `example.hello` procedure, generated by `create-t3-app`, which takes a Zod object containing a string `text` and never fails.

How should this input be specified in the workflow?
The examples imply that something like
```
...
trpc:
  query:
    example.hello: Matt
```
would be enough (though that feels wrong because how would it handle multiple input values?) but that doesn't work. tRPC reports that...
```
❌ tRPC failed on example.hello: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [],
    "message": "Required"
  }
]
```

Maybe something like the mutation example in the docs would work?
```
trpc:
  query:
     example.hello:
       text: Matt
```
But no, that causes the same error from tRPC.

I've also tried adopting syntax from the [HTTP API docs](https://docs.stepci.com/guides/testing-trpc.html) like
```
...
  json:
    text: Matt
```
```
...
  params:
    text: Matt
```
and
```
...
  data:
    text: Matt
```
I've even tried hand-writing the JSON object and sending that as a string but they all result in the same message from tRPC.

By the way, while these tests are failing to even send the data stepci is reporting that it passes no matter what status code I check against. Does anyone have experience testing tRPC with stepci and can show me where I'm going wrong?

**Comment by mishushakov** [comment]
Hello, thanks for starting the discussion 😄
The reason you're getting this error is because tRPC request batching is currently not supported by Step CI (as outlined in the documentation)

Please subscribe to the issue if you want to get updates: https://github.com/stepci/stepci/issues/106
Thank you so much for your effort 🙏

**Comment by mishushakov** [comment, marked as answer]
Hey again, @matthewd673
I've taken a closer look into the issue and have added tRPC batching support 🤩

The "Query Batching" example from the updated documentation now works with create-t3-app (I've tested it!)
https://docs.stepci.com/guides/testing-trpc.html

Hopefully, you're still keen to try Step CI ☺️
Please let me know if you run into any issues or have other questions I could help you with 😉

Mish & Team

**Reply by matthewd673** [reply]
Thanks for the quick response,
I'm still having issues using query batching. Using the new syntax in the docs
```
trpc:
  query:
    - example.hello:
        json:
          text: Matt
```
tRPC returns the following:
```
❌ tRPC failed on 0: No "query"-procedure on path "0"
```

When I swap that out with the old, non-batching syntax
```
trpc:
  query:
    example.hello: Matt
```
it obviously fails as before but returns
```
❌ tRPC failed on example.hello: [ ... (invalid type, etc.) ... ]
```
which indicates that the bad request is at least reaching the right place. It seems that stepci is sending a request to some path `0` when using query-batching instead of the expected `example.hello`. Is there anything I'm still doing wrong here?

Thanks again,
Matt

**Reply by mishushakov** [reply]
Interesting.
Can I see the full request URL? Do you have a trailing slash after the URL?

**Reply by mishushakov** [reply]
Btw. are you sure you're using the latest version, `2.6.4`?

**Comment by sebastianwis** [comment]
Hey Matthew @matthewd673
Thank you for your contribution here! 
It would be amazing to welcome you to our Discord if you haven't joined it yet: 
https://discord.com/invite/KqJJzJ3BTu
This is where we keep users up to date with updates, features and bug resolutions.
